src/flash.rs

```diff
@@ -34,3 +49,436 @@ impl ACR {
+        // 4. Check that the new number of wait states has been taken into account by reading back the FLASH_ACR register.
+    }
+
+    /// Unlock the flash memory, must only be executed on locked flash
+    fn unlock(&self, bank: Bank) -> Result<(), Error> {
+        const FLASH_KEY1: u32 = 0x4567_0123;
+        const FLASH_KEY2: u32 = 0xCDEF_89AB;
+
+        let regs = self.bank_registers(bank);
+        regs.keyr.write(|w| unsafe { w.bits(FLASH_KEY1) });
+        regs.keyr.write(|w| unsafe { w.bits(FLASH_KEY2) });
+
+        if regs.cr.read().lock().bit_is_clear() {
+            Ok(())
+        } else {
+            Err(Error::Unlock)
+        }
+    }
+
+    /// Lock the flash memory, must be executed after an unlock
+    fn lock(&mut self, bank: Bank) {
+        let regs = self.bank_registers(bank);
+        regs.cr.modify(|_, w| w.lock().set_bit());
+    }
+
+    /// Read flash memory from a given bank, sector and offset by the length of buff / 4
+    ///
+    /// # Panics
+    ///
+    /// Panics if the sector is beyond the number of sectors supported by the user bank for this device
+    /// Panics if the buff length is not a multiple of a 32bit word
+    /// Panics if the start offset is not a multiple of a 32bit word
+    /// Panics if the buff length exceeds the user bank 2 end address for this device
+    pub fn read_sector(
+        &mut self,
+        bank: Bank,
+        sector: usize,
+        start: usize,
+        buff: &mut [u8],
+    ) -> Result<(), Error> {
+        if sector > bank.num_sectors() {
+            panic!("Sector out of bounds")
+        }
+        if buff.len() % 4 != 0 || start % 4 != 0 {
+            panic!("buff length and start offset must be a multiple of a 32bit word");
+        }
+
+        let regs = self.bank_registers(bank);
+
+        // Check that no Flash main memory operation is ongoing by checking the BSY bit
+        while regs.sr.read().bsy().bit_is_set() {}
+
+        // H742 RM, section 4.3.8, Page 158
+        // Single read sequence
+        // The recommended simple read sequence is the following:
+        // 1. Freely perform read accesses to any AXI-mapped area.
+        // 2. The embedded Flash memory effectively executes the read operation from the read
+        // command queue buffer as soon as the non-volatile memory is ready and the previously
+        // requested operations on this specific bank have been served.
+
+        // Get the address of the sector/bank
+        let mut addr = bank.sector_address(sector) as *mut u32;
+
+        let mut count = 0;
+
+        unsafe {
+            // Offset it by the start position
+            addr = addr.add(start / 4);
+
+            let chunks = buff.chunks_exact_mut(4);
+
+            for chunk in chunks {
+                if addr as usize > bank.end_address() {
+                    if bank == Bank::UserBank2 {
```

Comment:
There's no reason to call read_sector recursively here. Perhaps you can remove that and simplify the code?

Reply:
Yeah, I wasn't sure if this was the best of ideas. I'll remove it 👍